### ISSUE TITLE
python tracing error in multiprocessing

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#1683]
Let's say there is a simple example.
```py
$ cat no-multi.py 
#!/usr/bin/env python3

def f(x):
    return x*x

input = [1, 2, 3]
output = list(map(f, input))
print(output)
```
Then it can simply be traced as follows.
```
$ uftrace no-multi.py 
[1, 4, 9]
# DURATION     TID     FUNCTION
            [ 24318] | __main__.<module>() {
   0.918 us [ 24318] |   f();
   0.314 us [ 24318] |   f();
   0.241 us [ 24318] |   f();
  11.788 us [ 24318] |   builtins.print();
  22.403 us [ 24318] | } /* __main__.<module> */
```
If it's changed to use multiprocessing, then it also works fine.
```py
$ cat multi.py 
#!/usr/bin/env python3

from multiprocessing import Pool

def f(x):
    return x * x

input = [1, 2, 3]
with Pool(5) as p:
    output = p.map(f, input)
    print(output)
```
```
$ ./multi.py 
[1, 4, 9]
```
But if uftrace is attached then it gets crashed as follows.
```
$ uftrace multi.py 
Traceback (most recent call last):
  File "/usr/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/honggyu/usr/lib/uftrace/uftrace.py", line 32, in <module>
    exec(code, new_globals)
  File "<string>", line 10, in <module>
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 367, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 540, in _handle_tasks
    put(task)
  File "/usr/lib/python3.10/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib/python3.10/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function f at 0x7fc25ba08550>: attribute lookup f on __main__ failed
# DURATION     TID     FUNCTION
            [ 24478] | __main__.<module>() {
  35.235 ms [ 24478] |   importlib._bootstrap._find_and_load();
   4.533 us [ 24478] |   importlib._bootstrap._handle_fromlist();
  31.624 ms [ 24478] |   multiprocessing.context.Pool();
   7.154 us [ 24478] |   multiprocessing.pool.__enter__();
 383.925 us [ 24478] |   multiprocessing.pool.map();
   1.890 ms [ 24478] |   multiprocessing.pool.__exit__();
  69.191 ms [ 24478] | } /* __main__.<module> */
```

**Comment by honggyukim** [comment]
The last error message `attribute lookup f on __main__ failed` looks suspicious because if the original `multi.py` is changed to use lambda then it also shows the similar error message even without uftrace.
```py
$ cat multi-lambda.py
#!/usr/bin/env python3

from multiprocessing import Pool

input = [1, 2, 3]
with Pool(5) as p:
    output = p.map(lambda x: x*x, input)
    print(output)
```
```
$ ./multi-lambda.py 
Traceback (most recent call last):
  File "/home/honggyu/work/uftrace/git/python/./multi2.py", line 7, in <module>
    output = p.map(lambda x: x*x, input)
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 367, in map
    return self._map_async(func, iterable, mapstar, chunksize).get()
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 774, in get
    raise self._value
  File "/usr/lib/python3.10/multiprocessing/pool.py", line 540, in _handle_tasks
    put(task)
  File "/usr/lib/python3.10/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib/python3.10/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
_pickle.PicklingError: Can't pickle <function <lambda> at 0x7fb9d2fe67a0>: attribute lookup <lambda> on __main__ failed
```

**Comment by honggyukim** [comment]
The test is done with the fixes we found at #1679.
```diff
diff --git a/cmds/record.c b/cmds/record.c
index 900eb48e..cab1503c 100644
--- a/cmds/record.c
+++ b/cmds/record.c
@@ -2232,6 +2232,12 @@ int do_child_exec(int ready, struct uftrace_opts *opts, int argc, char *argv[])
                python_path = strjoin(python_path, "misc", ":"); /* FIXME */
                setenv("PYTHONPATH", python_path, 1);
                free(python_path);
+
+               /*
+                * prevent from creating .pyc files inside __pycache__.
+                * it makes some script execution failed.
+                */
+               setenv("PYTHONDONTWRITEBYTECODE", "1", 1);
        }
 
        /*
diff --git a/python/uftrace.py b/python/uftrace.py
index 9683acc2..c0a21139 100644
--- a/python/uftrace.py
+++ b/python/uftrace.py
@@ -4,9 +4,12 @@ import sys
 sys.argv = sys.argv[1:len(sys.argv)]
 
 filename = sys.argv[0]
-if os.path.exists(filename) or filename.count('/') > 0:
+if os.path.exists(filename) or filename[0] == '/':
     os.environ["UFTRACE_PYMAIN"] = filename
-    pass
+    if filename[0] == '/':
+        pathname = filename
+    else:
+        pathname = os.getcwd() + '/' + filename
 else:
     for dir in os.environ["PATH"].split(":"):
         pathname = dir + '/' + filename
@@ -20,7 +23,11 @@ else:
             continue
 
 import uftrace_python
+new_globals = dict(globals())
+new_globals["__file__"] = pathname
+sys.path.insert(0, os.path.dirname(pathname))
+
 code = open(sys.argv[0]).read()
 sys.setprofile(uftrace_python.trace)
-exec(code)
+exec(code, new_globals)
 sys.setprofile(None)
```

**Comment by namhyung** [comment]
Hmm.. strange.  It works fine with the current master branch.  The change in uftrace.py broke it.

**Comment by honggyukim** [comment]
Ah. right. The current master is fine but the above made the problem.

I've found that the following fix makes all the issues gone.
```diff
diff --git a/python/uftrace.py b/python/uftrace.py
index d3941d76..f56c55d5 100644
--- a/python/uftrace.py
+++ b/python/uftrace.py
@@ -24,7 +24,7 @@ else:
             continue
 
 import uftrace_python
-new_globals = dict(globals())
+new_globals = globals()
 new_globals["__file__"] = pathname
 sys.path.insert(0, os.path.dirname(pathname))
 
```
```
$ uftrace multi.py 
[1, 4, 9]
# DURATION     TID     FUNCTION
            [ 31840] | __main__.<module>() {
  35.660 ms [ 31840] |   importlib._bootstrap._find_and_load();
   4.329 us [ 31840] |   importlib._bootstrap._handle_fromlist();
  31.241 ms [ 31840] |   multiprocessing.context.Pool();
   7.224 us [ 31840] |   multiprocessing.pool.__enter__();
            [ 31840] |   multiprocessing.pool.map() {
   3.219 us [ 31843] | f();
   3.127 us [ 31844] | f();
   8.887 us [ 31845] | f();
   1.042 ms [ 31840] |   } /* multiprocessing.pool.map */
  33.208 us [ 31840] |   builtins.print();
   1.629 ms [ 31840] |   multiprocessing.pool.__exit__();
  69.672 ms [ 31840] | } /* __main__.<module> */
```

**Comment by namhyung** [comment]
Thanks, I'll update the patch.

**Comment by namhyung** [comment]
Fixed by e26f74c83
